boltview/reduce.tcc

```diff
@@ -405,6 +439,85 @@ public:
 };
 #endif // __CUDACC__
 
+template<typename TOutputValue, int tInputDimension, int tReduceDimension, bool tRunOnDevice>
+class DimensionReduce {
+	static_assert(tReduceDimension < tInputDimension, "tReduceDimension must be less than tInputDimension");
+public:
+
+	template<typename TSize>
+	DimensionReduce(TSize size, ExecutionPolicy execution_policy = ExecutionPolicy{}) :
+		execution_policy_(execution_policy)
+	{}
+
+	template<typename TInView, typename TOutView, typename TOperator>
+	void runAsync(
+		TInView input, TOutView output,
+		TOutputValue initial_value, TOperator reduction_operator
+	) {
+		static_assert(TInView::kDimension == tInputDimension, "Invalid input dimension.");
+		static_assert(TOutView::kDimension == tInputDimension - 1, "Invalid output dimension.");
+
+		detail::ReduceImplementation<false>::run(
+			input, output, DimensionValue<tReduceDimension>(),
+			initial_value, reduction_operator, execution_policy_);
+	}
+
+	template<typename TInView, typename TOutView, typename TOperator>
+	void run(
+		TInView input, TOutView output,
+		TOutputValue initial_value, TOperator reduction_operator
+	) {
+		runAsync(input, output, initial_value, reduction_operator);
+	}
+
+	ExecutionPolicy execution_policy_;
+
+};
+
+#ifdef __CUDACC__
+
+template<typename TOutputValue, int tInputDimension, int tReduceDimension>
+class DimensionReduce<TOutputValue, tInputDimension, tReduceDimension, true> {
+	static_assert(tReduceDimension < tInputDimension, "tReduceDimension must be less than tInputDimension");
+public:
+	constexpr static int kBlockSize = 128;
+	constexpr static int kBucketSize = 2;
+
+	template<typename TSize>
+	DimensionReduce(TSize size, ExecutionPolicy execution_policy = ExecutionPolicy{}) :
+		execution_policy_(execution_policy),
+		tmp_image_(detail::estimateTemporaryBuffer(size, tReduceDimension, kBlockSize, kBucketSize))
+	{}
+
+	template<typename TInView, typename TOutView, typename TOperator>
+	void runAsync(
+		TInView input, TOutView output,
+		TOutputValue initial_value, TOperator reduction_operator
+	) {
+		static_assert(TInView::kDimension == tInputDimension, "Invalid input dimension.");
```

Comment:
We should probably check that the input.size() <= size from the constructor.

Reply:
sure - throw an exception (or at least assert)